evm-contracts/src/v0.7/dev/CompoundPriceFlaggingValidator.sol

```diff
@@ -0,0 +1,277 @@
+   * @param aggregators address[] memory
+   * @return address[] memory invalid aggregators
+   */
+  function update(address[] memory aggregators) public returns (address[] memory){
+    address[] memory invalidAggregators = check(aggregators);
+    s_flags.raiseFlags(invalidAggregators);
+    return invalidAggregators;
+  }
+
+  /**
+   * @notice Check the price deviation of an array of aggregators
+   * @dev If any of the aggregators provided have an equivalent Compound Oracle feed
+   * that with a price outside of the configured deviation, this function will return them.
+   * @param data bytes encoded address array
+   * @return needsUpkeep bool indicating whether upkeep needs to be performed
+   * @return invalid aggregators - bytes encoded address array of invalid aggregator addresses
+   */
+  function checkForUpkeep(bytes calldata data) external view override returns (bool, bytes memory) {
+    address[] memory invalidAggregators = check(abi.decode(data, (address[])));
+    bool needsUpkeep = (invalidAggregators.length > 0);
+    return (needsUpkeep, abi.encode(invalidAggregators));
+  }
+
+  /**
+   * @notice Check and raise flags for any aggregator that has an equivalent Compound
+   * Open Oracle feed with a price deviation exceeding the configured setting.
+   * @dev This contract must have write permissions on the Flags contract
+   * @param data bytes encoded address array
+   */
+  function performUpkeep(bytes calldata data) external override {
+    update(abi.decode(data, (address[])));
+  }
+
+  /**
+   * @notice Get the threshold of an aggregator
+   * @param aggregator address
+   * @return string Compound Oracle Symbol
+   * @return uint8 Compound Oracle Decimals
+   * @return uint32 Deviation Threshold Denominator
+   */
+  function threshold(address aggregator)
+    public
+    view
+    returns (string memory, uint8, uint32)
+  {
+    CompoundFeedDetails memory compDetails = s_thresholds[aggregator];
+    return(
+      compDetails.symbol,
+      compDetails.decimals,
+      compDetails.deviationThresholdDenominator
+    );
+  }
+
+  /**
+   * @notice Get the flags address
+   * @return address
+   */
+  function flags() external view returns (address) {
+    return address(s_flags);
+  }
+
+  /**
+   * @notice Get the Compound Open Oracle address
+   * @return address
+   */
+  function compoundOpenOracle() external view returns (address) {
+    return address(s_compOpenOracle);
+  }
+
+  /**
+   * @notice Check if an aggregator has an equivalent Compound Oracle feed
+   * that's price is deviated more than the threshold.
+   * @param aggregator address of the Chainlink aggregator
+   * @return invalid bool. True if the deviation exceeds threshold.
+   */
+  function isInvalid(address aggregator) private view returns (bool invalid) {
+    CompoundFeedDetails memory compDetails = s_thresholds[aggregator];
+    if (bytes(compDetails.symbol).length == 0) {
```

Comment:
I think checking that `compDetails.deviationThresholdDenominator == 0` would be a little cheaper but not sure.